swupd/manifest.go

```diff
@@ -342,72 +342,113 @@ func (m *Manifest) sortFilesVersionName() {
 	})
 }
 
-func (m *Manifest) addDeleted(oldManifest *Manifest) {
-	for _, df := range oldManifest.DeletedFiles {
-		if df.findFileNameInSlice(m.Files) == nil {
-			m.Files = append(m.Files, df)
-			m.DeletedFiles = append(m.DeletedFiles, df)
-		}
-	}
+func (m *Manifest) appendFile(f *File) {
+	m.Files = append(m.Files, f)
 }
 
 // linkPeersAndChange
 // At this point Manifest m should only have the files that were present
 // in the chroot for that manifest. Link delta peers with the oldManifest
-// if the file in the oldManifest is not deleted or ghosted
-func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) int {
-	changed := 0
-	for _, of := range oldManifest.Files {
-		if match := of.findFileNameInSlice(m.Files); match != nil {
+// if the file in the oldManifest is not deleted or ghosted.
+// Expects m and oldManifest files lists to be sorted by name only
+func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) (int, int, int) {
+	var unchanged, changed, removed, added []*File
+	i := 0
+	j := 0
+	mFilesLen := len(m.Files)
+	omFilesLen := len(oldManifest.Files)
+	for i < mFilesLen && j < omFilesLen {
+		nf := m.Files[i]
+		of := oldManifest.Files[j]
+		if nf.Name == of.Name {
+			// if it is the same name check if anything about the file has changed.
+			// if something has changed update the version to current version and
+			// record that the file has changed
 			if of.Status == statusDeleted || of.Status == statusGhosted {
+				// don't worry about ghosted or deleted files
+				i++
```

Comment:
I think we do need to worry here, extending added with nf. This allows for

1. Version 10
    1. fred - deleted
    1. fred.new - normal file
1. Version 20
     1. fred - which may be a rename of the previous fred.new

Reply:
ok